drgn_tools/iscsi.py

```diff
@@ -0,0 +1,149 @@
+# Copyright (c) 2024, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Helper for iscsi
+"""
+import argparse
+from typing import Iterator
+from typing import Tuple
+
+from drgn import Object
+from drgn import Program
+from drgn.helpers.common import escape_ascii_string
+from drgn.helpers.common.type import enum_type_to_class
+from drgn.helpers.linux.list import list_for_each_entry
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.scsi import for_each_scsi_host
+from drgn_tools.scsi import for_each_scsi_host_device
+from drgn_tools.scsi import get_scsi_device_name
+from drgn_tools.scsi import host_module_name
+from drgn_tools.table import print_dictionary
+from drgn_tools.util import enum_name_get
+
+ISCSI_SESSION_STATES = ["LOGGED_IN", "FAILED", "FREE"]
+
+
+def for_each_iscsi_host(prog: Program) -> Iterator[Object]:
+    """
+    Iterates through all scsi host devices and returns a
+    iterator of hosts backed by iscsi drivers
+    :returns: an iterator of ``struct Scsi_Host *``
+    """
+    for shost in for_each_scsi_host(prog):
+        if "iscsi" in host_module_name(shost).lower():
+            yield shost
+
+
+def for_each_iscsi_session(prog: Program) -> Iterator[Object]:
+    """
+    Iterates through all iscsi_cls_session and gets the associated iscsi_session
+    :returns: an iterator of ``struct iscsi_session *``
+
+    """
+    for cls_session in list_for_each_entry(
+        "struct iscsi_cls_session", prog["sesslist"].address_of_(), "sess_list"
+    ):
+        yield Object(
+            prog,
+            prog.type("struct iscsi_session", filename="libiscsi.h"),
+            address=cls_session.dd_data,
+        )
+
+
+def for_each_attached_device(
+    prog: Program, shost: Object
+) -> Iterator[Tuple[Object, str]]:
+    """
+    Iterates through all Scsi_Host and gets their associated scsi device and name
+    :returns: an iterator of (``struct scsi_device *``, ``str``)
+
+    """
+    for scsi_dev in for_each_scsi_host_device(prog, shost):
+        name = get_scsi_device_name(prog, scsi_dev)
+        yield scsi_dev, name
+
+
+def print_iscsi_report(prog: Program) -> None:
+    """
+    Prints a comprehensive report including iscsi table and session stats. More info can be added later if needed.
+    """
```

Comment:
Please add below code at the beginning of this function, this will make sure the helper bail out if iscsi is not used when invoking this helper directly from drgn shell while not corelen module.
```
    msg = ensure_debuginfo(prog, ["libiscsi", "scsi_transport_iscsi"])
    if msg:
        print(msg)
        return
```